File: chap4/Pseudo.c
// Ces programmes sont sous licence CeCILL-B V1.
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define MAX_LETTRES 200

// Dans cette fonction "lettre" est censé ne contenir 
// qu'un seul caractère.
// On vérifie si ce caractère apparaît dans la chaîne "mot"

bool appartient(char lettre, const char mot[])
{
   int i;
   bool resultat;
   resultat = false;
   for (i = 0; i <= (int)strlen(mot) - 1; i = i + 1) 
   {
      if (mot[i] == lettre)
      {
         resultat = true;
      }
   }
   return resultat;
}

int main()
{
   char pseudo[MAX_LETTRES];
   bool pseudoOk;
   int i;
   const char autorises[] = "abcdefghijklmnopqrstuvwxyz";
   pseudoOk = false;
   // On redemande un pseudo tant qu'il n'est pas correct 
   while (!pseudoOk)
   {
      printf("Entrer votre pseudo :\n");
      scanf("%s", pseudo);
      pseudoOk = true;
      // On vérifie que chaque caractère du pseudo est autorisé
      for (i = 0; i <= (int)strlen(pseudo) - 1; i = i + 1) 
      {
         if (!appartient(pseudo[i],autorises))
         {
            pseudoOk=false;
         }
      }
   }
}

File: chap4/Pseudo.ml
(* Ces programmes sont sous licence CeCILL-B V1. *)

let appartient lettre mot =
   (*
   Dans cette fonction "lettre" est censé ne contenir
   qu'un seul caractère.
   On vérifie si ce caractère apparaît dans la chaîne "mot"
   *)
   let resultat = ref false in
   for i = 0 to String.length mot - 1 do
     if mot.[i] = lettre then
       resultat := true
   done;
   !resultat

let _ =
  let autorises = "abcdefghijklmnopqrstuvwxyz" in
  let pseudoOk = ref false in
  (* On redemande un pseudo tant qu'il n'est pas correct *)
  while not !pseudoOk do
    print_string "Entrer votre pseudo :"; print_newline ();
    let pseudo = input_line stdin in
    pseudoOk := pseudo <> "";
   (* On vérifie que chaque caractère du pseudo est autorisé *)
    for i = 0 to String.length pseudo - 1 do
      if not (appartient pseudo.[i] autorises) then
        pseudoOk := false
    done
  done
